### DISCUSSION TITLE
Deleting objects leaves empty folders that cannot be deleted via the GUI/mc

### DISCUSSION THREAD

**Discussion by dwrolvink** [minio/minio#15257] [Q&A] [answered]
## Expected Behavior
Calling `mc rm <target>/<bucket>/folder/path -r --force` will remove all objects in that path AND remove that folder.

Same for selecting the folder in the GUI and running delete on it.

## Current Behavior
All objects are deleted (as far as I can tell) but all the empty folders remain

## Context
We use this bucket to temporarily store information that needs to be backed up. Having these empty folders remain will add up and the list will get longer and longer, which is mostly just annoying, not sure if it will have a performance impact or not.

## Regression

I have not found any other issues talking about this problem, nor can I find any docs on how one would remove (empty) folders.

## Your Environment
```
MinIO VERSION 
2022-06-10T16:59:15Z

mc version RELEASE.2021-05-26T19-19-26Z

stemcells:
- alias: default
  os: ubuntu-bionic
  version: 1.88
```

**Comment by harshavardhana** [comment]
Please upgrade your mc.

**Reply by dwrolvink** [reply]
Will try this, though the same action executed via the GUI has the same issue.

With the updated client: 
``` bash
$  mc2 --version
mc2 version RELEASE.2022-07-06T14-54-36Z (commit-id=81c4a5ad6ee47ff2fc264b9811d884984ae05f6e)
Runtime: go1.18.3 linux/amd64
Copyright (c) 2015-2022 MinIO, Inc.
License GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>

$  mc ls trash/bbr-droplets-bucket/2021_04_30_18_06_43
[2022-07-08 16:36:48 UTC]     0B droplets/  # this folder has an empty subfolder and so on a couple levels deep

$  mc rm trash/bbr-droplets-bucket/2021_04_30_18_06_43 -r --force
mc: <ERROR> No object/version found to be removed in `trash/bbr-droplets-bucket/2021_04_30_18_06_43`.

$  mc ls trash/bbr-droplets-bucket/2021_04_30_18_06_43
[2022-07-08 16:36:59 UTC]     0B droplets/
```

**Reply by harshavardhana** [reply]
this means that you have some dangling objects inside this folder that are not purged properly, they need to healed to be removed - perhaps an old state.

**Reply by dormanze** [reply]
I found that this situation does not exist in single mode.just appears in cluster mode.

**Reply by harshavardhana** [reply]
That's not possible @dormanze

**Reply by dormanze** [reply]
my bad.The version number of my cluster is [RELEASE.2021-04-18T19-26-29Z].I didn't find this problem in the latest version，You can update the version @dwrolvink

**Reply by dwrolvink** [reply]
I ran the following:

``` bash
$  mc admin heal --recursive --remove trash/bbr-droplets-bucket
 ◓  bbr-droplets-bucket/2022_07_09_20_0…7904a8f4c743ecd197470e0ee9b9f90cfb8
    0/4,073 objects; 499 GiB in 3m26s
    ┌────────┬───────┬─────────────────────┐
    │ Green  │ 3,732 │  91.5% ███████████  │
    │ Yellow │   345 │   8.5% ██           │
    │ Red    │     0 │   0.0%              │
    │ Grey   │     0 │   0.0%              │
    └────────┴───────┴─────────────────────┘

$  mc rm am2-acc/bbr-droplets-bucket -r --force --older-than 3d
```

And the folders are still there. I guess that's the yellow items causing that then.
The issue of having yellow items after a heal is a totally different question, and not very surprising as we recently had a cluster node that went down for some time, so this discussion can basically be closed.

Though I am really new to minio as a whole and especially the healing part, so if anyone has a link to a good resource regarding this subject then I'd very much appreciate it.

**Comment by dwrolvink** [comment, marked as answer]
This was the answer as far as I'm concerned, but I can select it because it is a sub-comment or something: https://github.com/minio/minio/discussions/15257#discussioncomment-3109625

**Comment by sandroden** [comment]
I had the same problem, it turns out I have versions enabled and the user doesn't have the possibility to see the check to delete also all versioned object.
I tried to guess how to give the user enough 'actions' to see the checkbox, but only giving `'s3:*'` worked.
Admin can in fact delete the folder completely. I think there's a bug as I'd guess theese permissions should be enough:

```
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:DeleteObject",
                "s3:DeleteObjectVersion",
                "s3:GetObject",
                "s3:GetObjectVersion",
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:PutObject"
            ],
            "Resource": [
                "arn:aws:s3:::wikidattica",
                "arn:aws:s3:::wikidattica/*"
            ]
        }
```
If you know that I'm missing an action or if you know that this behavior is intentional, please let me know.
